### PR TITLE
CI: quieten build output when under a CI

### PR DESCRIPTION
**Pull request by grahamwhaley**
When running under a CI, a lot of the console log output
comes from installing and building kata components, such
as the linux kernel.

When under the CI, `chronic` those commands to reduce the
log size.

Fixes: #993

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

### CODE REVIEW

**Comment by grahamwhaley** [comment]
Need to see how this flies in the CI...
/test

**Comment by grahamwhaley** [comment]
Strange, some travis failed with some `apt` type issues (I nudged one to rebuild), and the metrics is showing some errors that on the face of it appear unrelated:
```
22:14:45 Looking for baremetal cleanup script for arch x86_64
22:14:45 Running baremetal cleanup script for arch x86_64
22:14:45 INFO: kill stale process
22:14:45 /usr/share/defaults/kata-containers/configuration.toml: file /usr/share/kata-containers/kata-containers.img does not exist
22:14:45 FATAL: command execution failed
```

I'll check these out in the morning.

File: .ci/install_kata.sh [hunk 1]
```diff
@@ -9,24 +9,32 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+local_chronic() {
+	if [ -n "$CI" ]; then
```
Comment (chavafg) [inline .ci/install_kata.sh]: shouldn't this be `-z`

Comment (chavafg) [reply .ci/install_kata.sh]: or maybe check  for` [ "$CI" == true ]` ?

Comment (grahamwhaley) [reply .ci/install_kata.sh]: It should be `-z`, yep - mea culpa.

I did do a search to see if we had a common idiom to check the `CI` var:

```bash
$  git grep '"$CI"' | fgrep if
.ci/go-test.sh:         if [ "$CI" = true ] && [ -n "$KATA_DEV_MODE" ]; then
.ci/go-test.sh: if [ "$CI" = true ] || [ -n "$KATA_DEV_MODE" ]; then
.ci/install_cri_containerd.sh:if [ -n "$CI" ]; then
.ci/install_kata.sh:    if [ -n "$CI" ]; then
.ci/install_openshift.sh:if [ "$ID" != "fedora" ] && [ "$CI" == true ]; then
.ci/setup.sh:   if ! sudo docker version | grep -q "$docker_version" && [ "$CI" == true ]; then
.ci/setup.sh:           if [ "$CI" == true ] && grep -q "N" /sys/module/kvm_intel/parameters/nested 2>/dev/null; then
README.md:if [ "$CI" = true ]; then
integration/containerd/cri/integration-tests.sh:        if [ -n "$CI" ]; then
integration/openshift/run_openshift_tests.sh:if [ "$ID" != "fedora" ] && [ "$CI" == true ]; then
```
So, we have a mix of the two - but, I do think `true` is probably better - I just took the easy route... let me fix both of those things..

**Comment by grahamwhaley** [comment]
redone and pushed.
@chavafg - it also occurred to me overnight that we might want to use that code in other places, so I've pulled the func out to the `lib.sh` and called it `ci_chronic()`...

**Comment by chavafg** [comment]
cool, thanks @grahamwhaley

**Comment by jodh-intel** [comment]
lgtm.

... But the ICE man cometh! :smile:

```
18:19:50 INFO: Applying patch /home/jenkins/workspace/workspace/kata-metrics-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/packaging/kernel/patches/0001-NO-UPSTREAM-9P-always-use-cached-inode-to-fill-in-v9.patch
18:19:50 INFO: Applying patch /home/jenkins/workspace/workspace/kata-metrics-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/packaging/kernel/patches/0002-Compile-in-evged-always.patch
18:19:50 gcc: internal compiler error: Terminated (program cc1)
```

**Comment by grahamwhaley** [comment]
Ouch.
yeah, a **lot** of the CIs failed here. In fact, it feels like we have had a lot of sporadic CI fails this week that may not be 'our problem'... such as net timeouts and random timeouts etc.
I'm trawling my PRs right now (including the other long running chronic one), so I'll have a snoot at the logs and see what I find....

**Comment by jodh-intel** [comment]
`clang` anyone? :)

/me goes to check the `gcc` manual for `--no-crash --jfdi` options...

**Comment by grahamwhaley** [comment]
Most CIs failed, in different ways. Some aborted and I can't even tell why?? And then one CI passed, but still dumped it mega-monster logs out... let's try a complete CI respin - I think maybe a comet passed overhead and threw some cosmic rays into the machinery whilst this PR was being tested....
/test

**Comment by grahamwhaley** [comment]
still unhappy CIs - I'll debug it....

**Comment by grahamwhaley** [comment]
OK, we were getting build timeouts as the kernel build at least can take more than the 5 minute CI timeout.
I've cloned in the chronic hack from #810. A couple of notes then:

- I've cloned in the chronic ticker code from #810 right now - depending on how these two PRs race, one of them can land the script and the other can then use it... for now I've hacked it in here so we can at least test it.
- when I ran this locally I see:

``` CI=true ./install_kata.sh
Install kata-containers image
.
Install Kata Containers Kernel
./install_kata.sh: line 20: 11856 Terminated              ( local sleepval=10; local echoval=60; local count=0; while true; do
    printf "."; sleep ${sleepval}; ((count+=${sleepval})); printf $count; ((count%${echoval} == 0)) && printf ":\n" && count=0;
done )
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.20.30.40.50.60:
.10.
Install Qemu
.10../lib.sh: line 60: 11982 Terminated              ( local sleepval=10; local echoval=60; local count=0; while true; do
    printf "."; sleep ${sleepval}; ((count+=${sleepval})); printf $count; ((count%${echoval} == 0)) && printf ":\n" && count=0;
done )
... and so on
```

@jodh-intel @marcov - any ide where the 'Terminated' line is coming from, and how I suppress it??

/test

**Comment by grahamwhaley** [comment]
oh, sigh... that metrics fail is way out of wack on boot times, which should be completely unrelated to this PR:

```
16:11:52 time="2018-12-13T16:11:52Z" level=warning msg="Failed Maxval (3.290250 < 4.147650) for [boot-times]"
16:11:52 time="2018-12-13T16:11:52Z" level=warning msg="Check for [boot-times] failed [Failed]"
16:11:52 time="2018-12-13T16:11:52Z" level=warning msg=" with [[*F* boot-times 93.0% 134.9% 107.0% 14.0% 132.3% 160.1% 21.0% 4.3% 20]]"
16:11:52 time="2018-12-13T16:11:52Z" level=warning msg="Overall we failed"
16:11:52 
16:11:52 Report Summary:
16:11:52 +-----+----------------------+-------+--------+--------+-------+--------+--------+-------+------+-----+
16:11:52 | P/F |         NAME         |  FLR  |  MEAN  |  CEIL  |  GAP  |  MIN   |  MAX   |  RNG  | COV  | ITS |
16:11:52 +-----+----------------------+-------+--------+--------+-------+--------+--------+-------+------+-----+
16:11:52 | *F* | boot-times           | 93.0% | 134.9% | 107.0% | 14.0% | 132.3% | 160.1% | 21.0% | 4.3% |  20 |
16:11:52 | P   | memory-footprint     | 85.0% | 106.1% | 115.0% | 30.0% | 106.1% | 106.1% | 0.0%  | 0.0% |   1 |
16:11:52 | P   | memory-footprint-ksm | 95.0% | 97.5%  | 105.0% | 10.0% | 97.5%  | 97.5%  | 0.0%  | 0.0% |   1 |
16:11:52 +-----+----------------------+-------+--------+--------+-------+--------+--------+-------+------+-----+
```

I don't immediately see any reason why... will have to keep another (my 3rd?) eye on the metrics again...

File: .ci/lib.sh [hunk 2]
```diff
@@ -87,7 +126,7 @@ function install_yq() {
 	GOPATH=${GOPATH:-${HOME}/go}
 	local yq_path="${GOPATH}/bin/yq"
 	local yq_pkg="github.com/mikefarah/yq"
-	[ -x  "${GOPATH}/bin/yq" ] && return
+	[ -x "${GOPATH}/bin/yq" ] && return
```
Comment (marcov) [inline .ci/lib.sh]: this can be changed in `[ -x "$yq_path" ] && return`

File: .ci/lib.sh [hunk 3]
```diff
@@ -30,6 +30,45 @@ info() {
...
+				((count%${echoval} == 0)) && printf ":\n" && count=0
+			done
+		)&
+
+		local printerPid="$!"
+
+		wait "$cmdPid"
+		local ret=$?
+		printf "\n"
+		kill "$printerPid"
```
Comment (marcov) [inline .ci/lib.sh]: it's me again :) 
So after some online research I figured out that you could add a `wait "$printerPid"` right after the `kill "$printerPid"` to silence the `Terminated` message.

**Comment by grahamwhaley** [comment]
Thanks @marcov 
I added that `yq_path` fix as a simple extra commit.
I've slid in the `wait`, and let's see what the CIs think...
/test

**Comment by grahamwhaley** [comment]
Jenkins was not responding - let's try now...
/test

**Comment by jodh-intel** [comment]
Nada.

/retest

**Comment by jodh-intel** [comment]
It's alive!

**Comment by grahamwhaley** [comment]
@jodh-intel - are you sure it triggered OK - I only see 5 CI entries here...
@chavafg - keep an eye open for CI triggering OK - since it was reboot/started this morning I'm not sure if it is fully running yet...
I'm gonna wait for input before trying to re-trigger here again..

**Comment by jodh-intel** [comment]
@grahamwhaley - yep, it triggered and showed a bunch of pending builds.

File: .ci/lib.sh [hunk 4]
```diff
@@ -30,6 +30,49 @@ info() {
...
+		# Wait for it to die, which silences the 'Terminated' text
+		# we'd otherwise get printed, which messes up the logs.
+		wait "$pinterPid"
```
Comment (chavafg) [inline .ci/lib.sh]: nit: `s/pinterPid/printerPid`
```
.ci/lib.sh: line 66: pinterPid: unbound variable
```

**Comment by chavafg** [comment]
looks weird as only few jobs are reporting here. I think that we can retrigger (after above nit is fixed) and see how it goes

**Comment by grahamwhaley** [comment]
grrr, typos! thanks @chavafg re-pushed....
/test

**Comment by grahamwhaley** [comment]
and to confirm, I see a bunch of CI's scheduled ....

**Comment by ydjainopensource** [comment]
I think the builds are failing due to a ci issue.

**Comment by grahamwhaley** [comment]
@ydjainopensource @marcov - looks like we've still not quite got the chronic background ticker kill stuff right - I need to revisit this again soon. From 16.04 logs:

```
Install kata-containers image
.10.20.30.40.50.60:
.10.
.ci/lib.sh: line 66: 11040 Terminated              ( local sleepval=10; local echoval=60; local count=0; while true; do
    printf "."; sleep ${sleepval}; ((count+=${sleepval})); printf $count; ((count%${echoval} == 0)) && printf ":\n" && count=0;
done )
Build step 'Execute shell' marked build as failure
```

**Comment by marcov** [comment]
@grahamwhaley given bash limitations, we could write a `gochronic` tool that does what we need.

**Comment by grahamwhaley** [comment]
@marcov :-) very tempting to try and upstream a 'heartbeat ticker' option to `chronic` itself ;-) I did have a quick look around to see if there were any obvious alternatives....
Let me see if I get to have a look at all of this again tomorrow..

**Comment by marcov** [comment]
@grahamwhaley let me know how it goes. I had a quick look at it, but Perl is not my thing :) 

Given the trivial task a chronic clone in go should be easy to do, I'm considering it as a pet project.

**Comment by grahamwhaley** [comment]
@marcov :-) heh heh. First I might try one more shell hack, and see if we can catch the signal inside the spin loop and die 'nicely', and see if that helps....

**Comment by jodh-intel** [comment]
:camel: or :goat: :fog: ... ?

That's "goat fog" as I couldn't find `:gopher:` :smile:

/me wishes github would show bubble help with the alt values. sigh...

**Comment by jodh-intel** [comment]
... oh, or :shell:

**Comment by grahamwhaley** [comment]
one last try at the shell then....
/test

**Comment by jodh-intel** [comment]
Another gnarly one - do we continue or kill it?

**Comment by grahamwhaley** [comment]
Indeed - the simplest looking quieten/chronic items turned out to clash majorly with the CI.
I'll try to visit these two this week - and if I can't find a suitable way to trivially fix them, I'll shut them down and update the Issues to note what we found etc.

**Comment by grahamwhaley** [comment]
Oh, I remember this one now - this is the PR that took the CI out, as it hung up in a loop and basically consumed all of the build slaves.... so, I've put a DNM on this to either figure out why it got stuck in a loop, or to close the PR....

**Comment by jodh-intel** [comment]
Any update @grahamwhaley ? Branch is now conflicted.

**Comment by grahamwhaley** [comment]
Realistically I'm not going to get to look at quietening the build logs any time soon again. Closing this PR ...

**Pull request closed without merging**